pulsar-broker/src/main/java/org/apache/pulsar/broker/service/DrainingHashesTracker.java

```diff
@@ -231,13 +300,14 @@ public synchronized void endBatch() {
      *
      * @param consumer   the consumer
      * @param stickyHash the sticky hash
-     * @param closing
+     * @param closing    whether the consumer is closing
      */
-    public synchronized void reduceRefCount(Consumer consumer, int stickyHash, boolean closing) {
+    public void reduceRefCount(Consumer consumer, int stickyHash, boolean closing) {
         if (stickyHash == 0) {
             return;
         }
-        DrainingHashEntry entry = drainingHashes.get(stickyHash);
+
+        DrainingHashEntry entry = getEntry(stickyHash);
```

Comment:
is this thread-safe? should we put this after the lock is acquired?

Reply:
Yes. getEntry method uses a read lock. It's intentional that the lock lifetime is as short as possible and that no other locks are acquired while holding the lock to reduce the chances for deadlocks. That's the strategy of this PR.